src/ClientData/include/ClientData/FastRenderingUtils.h

```diff
@@ -9,17 +9,39 @@
 
 namespace orbit_client_data {
 
+// Free Function that returns the number of the pixel of a particular timestamp. The query
+// assumes a closed-open interval [start_ns, end_ns), so end_ns is not a visible timestamp.
+[[nodiscard]] inline uint64_t GetPixelNumber(uint64_t timestamp_ns, uint32_t resolution,
+                                             uint64_t start_ns, uint64_t end_ns) {
+  uint64_t current_ns_from_start = timestamp_ns - start_ns;
+  uint64_t total_ns = end_ns - start_ns;
+
+  // Given a resolution of 4000 pixels, we can capture for 53 days without overflowing.
+  return (current_ns_from_start * resolution) / total_ns;
+}
+
 // Free Function that will be used for any implementation of
 // TimerDataInterface::GetTimersAtDepthDiscretized to get the next pixel timestamp. The query
 // assumes a closed-open interval [start_ns, end_ns), so end_ns is not a visible timestamp.
-[[nodiscard]] uint64_t GetNextPixelBoundaryTimeNs(uint64_t current_timestamp_ns,
-                                                  uint32_t resolution, uint64_t start_ns,
-                                                  uint64_t end_ns);
+[[nodiscard]] inline uint64_t GetNextPixelBoundaryTimeNs(uint64_t current_timestamp_ns,
+                                                         uint32_t resolution, uint64_t start_ns,
+                                                         uint64_t end_ns) {
+  uint64_t current_pixel = GetPixelNumber(current_timestamp_ns, resolution, start_ns, end_ns);
+  uint64_t next_pixel = current_pixel + 1;
 
-// Free Function that returns the number of the pixel of a particular timestamp. The query
-// assumes a closed-open interval [start_ns, end_ns), so end_ns is not a visible timestamp.
-[[nodiscard]] uint64_t GetPixelNumber(uint64_t timestamp_ns, uint32_t resolution, uint64_t start_ns,
-                                      uint64_t end_ns);
+  // Calculates `ceil(dividend / divisor)` only using integers assuming dividend and divisor are not
+  // 0.
+  const auto rounding_up_division = [](uint64_t dividend, uint64_t divisor) -> uint64_t {
```

Comment:
Since you are concerened about the performance of these functions: I would get rid of this lambda. I'm not sure how smart compilers are / can be with inlining this; it might generate a call below.

Reply:
Yeah, that's what I meant. lgtm!